api/content.go

```diff
@@ -0,0 +1,59 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/ourrootsorg/cms-server/model"
+	"gocloud.dev/blob"
+)
+
+// ContentRequest contains a content type of the content to post
+type ContentRequest struct {
+	ContentType string `json:"contentType"`
+}
+
+// ContentResult contains a URL to post the content
+type ContentResult struct {
+	PutURL string `json:"putURL"`
+	Key    string `json:"key"`
+}
+
+// PostContentRequest returns a URL for posting content
+func (api API) PostContentRequest(ctx context.Context, contentRequest ContentRequest) (*ContentResult, *model.Errors) {
+	bucket, err := api.OpenBucket(ctx)
+	if err != nil {
+		return nil, model.NewErrors(http.StatusInternalServerError, err)
+	}
+	defer bucket.Close()
+
```

Comment:
I like "api". I've called it "routes" in the past, but I think "api" is just as good if not better.

Here's what I'm thinking we'd do for uploading content: 
1. client requests a URI to upload the content and gets back a PUT URI and a "key" (path)
2. client PUTs the content at the URI
3. client POSTs to /posts with a JSON that includes metadata about the post and the key of the content uploaded.
4. server saves the post metadata with a record-status of "pending", and sends a message to a queue
5. a "record-writer" process listens to the queue (lambda trigger or rabbitmq), reads the uploaded content, writes the records to the database, and updates the "post" metadata to say that the post record-status is now "ready".

I wasn't sure where to put the blob and pubsub init functions to be honest. I was thinking that the record-writer process would be a sibling directory to server. It needs access to the blob and pubsub services to access the uploaded data and the queue. In addition, the record-writer process also needs to be able to write records and update the post in the database. So, the record-writer process either:

a. creates a PostgresPersister and writes directly to the PostPersister interface to update the database, or

b. creates an API object, initializes it with the Persisters it's going to use, and uses API functions to update the database. 

If we go with option (b), then it would make sense for the blob and pubsub init functions to be in the "api" directory, probably directly in api.go, and for the record-writer process to call the blob and pubsub functions of the api object.

If we go with option (a) then it seems to make sense for the blob and pubsub init functions to be separate from the "api" directory, either in a "services" directory or in the "persist" directory, and to ask the records-writer to create its own instances of the blob and pubsub services.

Which level do you think the records-writer process should use? (We'll likely have other processes that listen to queues and process data, so what we decide now would ideally be the decision for other async processes as well.)

Reply:
I think the underlying question is, is the "api" directory something that is server-specific, or is a a set of reusable functions that any process could (and should) use?

Reply:
I'm used to thinking of microservices as owning their own database (where a "database" is a set of related tables, more a schema than a physical database). So in that model, if one service needs to put data into a table owned by another service it must call that service (or put a message on a queue for it). In that model, the record-writer process would use a REST call or a message to the server to write records and update the post in the database. I'm not saying we should do it that way here, just that that's the way I'm used to doing this sort of thing. I've got to finish this comment now. Perhaps we should talk it through tomorrow?

Reply:
We'd need to figure out which process owns the "records" table: server or records-writer. The records-writer process might want to insert 50,000 rows into the table, and if the server owns the table, then it would need to either make 50,000 calls to the server, batch up the records into groups of N each for the server to process without going over the payload or time limits, or insert the records itself. But if the records-writer process inserts the records and manages the records table, what happens when the user wants to read the records? The records-writer is an async process, so I don't know how it would respond. 

As far as I can tell, Lambda functions have a timeout of 15 minutes, but HTTP API endpoints in the AWS Gateway have a timeout of only 30 seconds. That's why I was thinking that the long-running functions would be handled by async lambda functions.

In general I agree with you that each service should own its own database tables, but I'm not sure how to handle this case. I'm open to suggestions. Let's talk tomorrow.